lib/codelab.dart

```diff
@@ -97,13 +115,38 @@ class CodelabUi {
     dialog = Dialog();
   }
 
+  void _initBusyLights() {
+    busyLight = DBusyLight(querySelector('#dartbusy'));
+  }
+
   void _initEditor() {
     // Set up CodeMirror
     editor = (editorFactory as CodeMirrorFactory)
         .createFromElement(_editorHost, options: codeMirrorOptions)
           ..theme = 'darkpad'
           ..mode = 'dart'
           ..showLineNumbers = true;
+
+    sourceProvider = CodelabDartSourceProvider(editor);
+    docHandler = DocHandler(editor, sourceProvider);
+
+    editor.document.onChange.listen((_) => busyLight.on());
+    editor.document.onChange
+        .debounce(Duration(milliseconds: 1250))
+        .listen((_) => _performAnalysis());
+
+    editorFactory.registerCompleter(
+        'dart', DartCompleter(dartServices, editor.document));
+
+    // Listen for changes that would effect the documentation panel.
+    editor.onMouseDown.listen((e) {
+      // Delay to give codemirror time to process the mouse event.
```

Comment:
CodeMirror is quite the beast. It does a fair amount of DOM manipulation to achieve something that approximates an editor in a browser, IIUC.

Reply:
This code is copy pasted from playground.dart so I'm not sure, but I think this is a safeguard against getting the previous cursor position instead of the new one when `_cursorPositionIsWhitespace()` is called

Reply:
I'll de-dupe this code, though.